### ISSUE TITLE
Should config passed through `--config` be authoritative?

### ISSUE THREAD

**Issue by Pothulapati** [linkerd/linkerd2#5158]
## Bug Report

Currently, Whatever configurations passed through `--config` with upgrades act as mutations to the current state which means there will be a difference in what `--config` passed during an upgrade and what the current state is (because there could've been previous mutations).

One problem that I came across is when I want to replace all the existing config with a new config passed through `--config`. My initial suggestion was to have a similar flag like `--addon-overwrite` i.e `--overwrite` which skips retrieving existing configuration and replaces with the default + overwrite config passed through `--config`.

But based on my conversations with @olix0r, It would be better to have config passed through `--config` as authoritative and use it without mutating existing config.

### The Helm way

Obviously, It felt better to understand how Helm does this, and maybe follow a simialr methodology here. Helm, when there is no new config passed (through `--set`, `-f`), re-uses the existing config when a upgrade is performed. (This is the expected case).

But when there is extra config specified through `--set`, `--f`, It **skips retrieving the current config and considers whatever passed currently as authoritative unless `--reuse-values` flag is specified**

More on this can be found [here](https://medium.com/@kcatstack/understand-helm-upgrade-flags-reset-values-reuse-values-6e58ac8f127e)

Should we follow a similar approach to have `--config` as authoritative by default and allow mutations only when `--reuse-config` is configured?

@adleong @olix0r

**Comment by adleong** [comment]
Just to be clear, this question is specifically about `linkerd upgrade`, right?  i.e. this is not relevant to `linkerd install` and not relevant to any linkerd extensions (since those do not have upgrade workflows).

It strikes me as very odd that Helm defaults to "reset-values" mode when the `--set` flag is used.  Here is what I would intuitively expect Linkerd to do:

* When `linkerd upgrade` is run without flags, the existing configuration is reused
* When `linkerd upgrade` is run with configuration flags (such as `--disable-heartbeat` for example), those mutations are applied on top of the existing configuration.  This is the "reuse-values" behavior of helm.
* Similarly, when `linkerd upgrade` is used with the `--set` flag, those mutations should be applied on top of the existing configuration.  Essentially, the configuration flags are a shorthand for `--set`.
* When `linkerd upgrade` is run with the `--values` flag (previously named `--config`), the existing configuration is discarded and the given config is applied on top of the chart defaults.  This is the "reset-values" behavior of helm.
* When `linkerd upgrade` is run with both `--values` and `--set`, the give config is applied on top of the chart defaults, and then the `--set` mutations are applied on top of that.

In other words, I would expect `linkerd upgrade` to "reset-values" if and only if the `--values` flag is used.

**Comment by adleong** [comment]
Actually, I think it is probably simpler and more coherent to just ALWAYS do the "reuse-values" behavior, even if the `--values` flag is set.  This means that there's no way in the CLI to unset a value, but I think that's a pretty niche case.  (If this is really necessary, users can unset values by manually editting the linkerd-config-overrides secret). 

I think the "reset-values" behavior isn't that useful here because it defeats the entire purpose of the upgrade command.  If you want to provide the fresh set of values yourself, you should really do something like:

```
linkerd install --ignore-cluster --values values.yml | kubectl apply --prune -l linkerd.io/control-plane-ns=linkerd -f -
```

**Comment by olix0r** [comment]
> If you want to provide the fresh set of values yourself, you should really do something like: ...

Though, this necessarily requires downtime, as reinstalling like this builds new trust roots...

Might we be able to support an `--unset` flag?

**Comment by adleong** [comment]
The trust roots are in the values so as long as you give the same trust roots in your values.yml, it shouldn't incur any downtime.

**Comment by olix0r** [comment]
Oh, I see what you're saying:

> In other words, I would expect linkerd upgrade to "reset-values" if and only if the --values flag is used.

+1

**Comment by adleong** [comment]
This has been addressed in https://github.com/linkerd/linkerd2/pull/5507.  The behavior is that described in https://github.com/linkerd/linkerd2/issues/5158#issuecomment-758306503: `linkerd upgrade` always emulates the behavior of Helm's "reuse-value" mode.
